### PR TITLE
WIP: Parse CSV directly from input stream 

### PR DESCRIPTION
**Pull request by BaurzhanSakhariev**
Inspired by https://github.com/FasterXML/jackson-docs/wiki/Presentation:-Jackson-Performance
>Use "unrefined" (least processed) forms of input: i.e., do not try decorating input sources and output targets:
Input: byte[] is best if you have it; InputStream second best; followed by Reader -- and in every case, do NOT try reading input into a String!

We do use `byte[]` in `CSVLineParser` which is the best option according to the docs, however, in our case it's a result of a roundtrip [read string](https://github.com/crate/crate/blob/master/server/src/main/java/io/crate/execution/engine/collect/files/FileReadingIterator.java#L248) --> [get bytes](https://github.com/crate/crate/blob/master/server/src/main/java/io/crate/operation/collect/files/CSVLineParser.java#L88). 

We can utilise Jackson lib's methods to consume `InputStream` directly as suggested in their docs ^ and thus avoid encode/decode string roundtrip.

This is not how final implementation should look like, more structural changes might be needed due to the switch of responsibility of some components - deepest component in the pipeline (parser) now gets data on its own (also should take care of line numbers on its own) and top level `FileReadingIterator` only creates/closes `InputStream`.

For the sake of simplicity and solely to understand how this change affects overall performance only happy path of some limited scenario is tested - no retries, no compression, no headers, csv file only, no line numbers in summary. Some is easy already (compression, line numbers but omitted to keep it simple), retries and headers might need more changes, skipped for now.

### CODE REVIEW

File: server/src/main/java/io/crate/execution/engine/collect/files/FileReadingIterator.java [hunk 1]
```diff
@@ -226,59 +222,26 @@ private boolean shouldBeReadByCurrentNode(URI uri) {
...
-                if (currentLineNumber < startFrom) {
-                    continue;
-                }
-                if (line.length() == 0) {
-                    continue;
-                }
-                break;
-            }
-        } catch (SocketTimeoutException e) {
-            if (retry > MAX_SOCKET_TIMEOUT_RETRIES) {
```
Comment (BaurzhanSakhariev) [inline server/src/main/java/io/crate/execution/engine/collect/files/FileReadingIterator.java:259]: component which actually reads the data has to take care of retries - so need to push this logic further down the pipeline. for now only happy path is tested

File: server/src/main/java/io/crate/execution/engine/collect/files/LineParser.java [hunk 2]
```diff
@@ -50,25 +54,29 @@ private enum InputType {
...
         }
     }
 
-    public byte[] getByteArray(String line, long rowNumber) throws IOException {
+    @Nullable
+    public byte[] getByteArray() throws IOException {
         if (inputType == InputType.CSV) {
-            return parserProperties.fileHeader() ?
-                csvLineParser.parse(line, rowNumber) : csvLineParser.parseWithoutHeader(line, rowNumber);
+            return csvLineParser.parse(); // this currently does things similar to parseWithoutHeader
```
Comment (BaurzhanSakhariev) [inline server/src/main/java/io/crate/execution/engine/collect/files/LineParser.java:73]: we would test only CSV without header in the current iteration

File: server/src/main/java/io/crate/operation/collect/files/CSVLineParser.java [hunk 3]
```diff
@@ -84,6 +104,43 @@ public void parseHeader(String header) throws IOException {
         }
     }
 
+    @Nullable
+    public byte[] parse() throws IOException {
+
+        if (parser.getInputSource() == null) {
```
Comment (BaurzhanSakhariev) [inline server/src/main/java/io/crate/operation/collect/files/CSVLineParser.java:110]: had to add this, despite on 
csvReader = csvReader.without(JsonParser.Feature.AUTO_CLOSE_SOURCE);

parser closes stream when file is read up - that's why commented out it above. 
it's actually fine behavior, we want to close when file is exhausted but would be better to control it explicitly. 
closeCurrentStream probably not needed now (falls to == null anyway). 
To sum up - it's fine only for testing perf as it doesn't break correctness, file is read up, resource are released but we need to control it explicitly in the real implementation i.e make `csvReader.without(JsonParser.Feature.AUTO_CLOSE_SOURCE);` work

File: plugins/cr8-copy-s3/src/test/java/io/crate/copy/s3/S3FileReadingCollectorTest.java [hunk 4]
```diff
@@ -1,256 +1,256 @@
...
-                        when(objectListing.isTruncated()).thenReturn(false);
-                        return client;
-                    }
-                }, uri, "https")),
-            false,
-            1,
-            0,
-            List.of("id", "name", "details"),
-            CopyFromParserProperties.DEFAULT,
-            FileUriCollectPhase.InputFormat.JSON,
```
Comment (BaurzhanSakhariev) [inline plugins/cr8-copy-s3/src/test/java/io/crate/copy/s3/S3FileReadingCollectorTest.java:243]: this is for JSON which is stubbed in this PR.

**Comment by BaurzhanSakhariev** [comment]
tests: I removed unit tests and left only relevant integration tests which are passing. Adapting unit tests to compile/pass is out of the scope of the goal in the first iteration - make `COPY FROM `with some specific config - no-header, no gzip

**Comment by BaurzhanSakhariev** [comment]
```
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |      473.789 ±    0.000 |    473.789 |    473.789 |    473.789 |    473.789 |
|   V2    |      416.715 ±    0.000 |    416.715 |    416.715 |    416.715 |    416.715 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               -  12.82%                           -  12.82%   

```

**Comment by BaurzhanSakhariev** [comment]
only 12% gain and for integrating this change we might need to restructure a lot of code - I think it doesn't worth it

**Comment by matriv** [comment]
Imho, 12% is not negligible, but let's hear other opinions as well.
Also, If it's easy, could you try different number of records, like 10k, 100k, 1m, 10m, (100m)
to see how it affects when the #records increases?

**Comment by BaurzhanSakhariev** [comment]
there is another related [change](https://github.com/crate/crate/pull/12795#issuecomment-1193996058) which gains only 5% but requires some not nice changes (like propagating one instance through the chain of classes since currently top level `FileReadingIterator` is responsible for the resources. If we ever go for re-structure we might combine those optimisation and maybe improve 12 further.

**Pull request closed without merging**